File: src/validate.ts
/**
 * Проверка сохранности: правка прозы не должна портить защищённое
 * содержимое. Сравнивает исходник и исправленный текст и сообщает,
 * что пропало или изменилось.
 *
 * Разрешённые скиллом изменения не считаются нарушением: смена прямых
 * кавычек на «ёлочки», десятичной точки на запятую, Title Case в
 * заголовке на обычный регистр, удаление параметра utm_source ИИ-инструмента.
 */

import { analyze } from "./detect.ts";
import type { ContextMode } from "./types.ts";

export interface Violation {
  kind: string;
  detail: string;
}

export interface ValidationResult {
  ok: boolean;
  violations: Violation[];
  issuesBefore: number;
  issuesAfter: number;
}

function all(re: RegExp, s: string): string[] {
  return [...s.matchAll(re)].map((m) => m[0]);
}

function frontmatter(s: string): string {
  return /^---\n[\s\S]*?\n---(?:\n|$)/.exec(s)?.[0] ?? "";
}

function stripCode(s: string): string {
  return s.replace(/^(```|~~~)[^\n]*\n[\s\S]*?^\1[^\n]*$/gm, "").replace(/`[^`\n]+`/g, "");
}

const AI_PARAM = /[?&](?:utm_source=(?:chatgpt\.com|openai|copilot\.com|claude\.ai|perplexity\.ai|perplexity|gemini\.google\.com|deepseek\.com)|referrer=grok\.com)/gi;

function urls(s: string): string[] {
  return all(/https?:\/\/[^\s)>\]»"]+/g, stripCode(s)).map((u) => u.replace(AI_PARAM, "").replace(/[?&]$/, ""));
}

function numbers(s: string): string[] {
  const prose = stripCode(s).replace(/https?:\/\/\S+/g, "");
  return all(/\d+(?:[.,]\d+)*/g, prose).map((n) => n.replace(/,/g, "."));
}

function multisetDiff(a: string[], b: string[]): { missing: string[]; added: string[] } {
  const count = new Map<string, number>();
  for (const x of a) count.set(x, (count.get(x) ?? 0) + 1);
  const added: string[] = [];
  for (const x of b) {
    const c = count.get(x) ?? 0;
    if (c > 0) count.set(x, c - 1);
    else added.push(x);
  }
  const missing: string[] = [];
  for (const [x, c] of count) for (let i = 0; i < c; i += 1) missing.push(x);
  return { missing, added };
}

function compareExact(kind: string, before: string[], after: string[], out: Violation[]): void {
  const { missing, added } = multisetDiff(before, after);
  for (const m of missing.slice(0, 5)) out.push({ kind, detail: `пропало или изменено: ${m.slice(0, 120)}` });
  for (const a of added.slice(0, 5)) out.push({ kind, detail: `появилось: ${a.slice(0, 120)}` });
}

function headingShape(s: string): string[] {
  return all(/^#{1,6}(?=\s)/gm, stripCode(s));
}

export function validate(before: string, after: string, context: ContextMode = "general"): ValidationResult {
  const v: Violation[] = [];
  if (frontmatter(before) !== frontmatter(after)) v.push({ kind: "yaml", detail: "YAML-шапка изменена" });
  compareExact("код", all(/^(```|~~~)[^\n]*\n[\s\S]*?^\1[^\n]*$/gm, before), all(/^(```|~~~)[^\n]*\n[\s\S]*?^\1[^\n]*$/gm, after), v);
  compareExact("инлайн-код", all(/`[^`\n]+`/g, stripCode(before)), all(/`[^`\n]+`/g, stripCode(after)), v);
  compareExact("формула", all(/\$\$[\s\S]*?\$\$|\$[^$\n]+\$/g, stripCode(before)), all(/\$\$[\s\S]*?\$\$|\$[^$\n]+\$/g, stripCode(after)), v);
  compareExact("URL", urls(before), urls(after), v);
  compareExact("ссылка на литературу", all(/\[\d+(?:[,;–-]\s*\d+)*(?:,\s*с\.\s*\d+(?:[–-]\d+)?)?\]/g, stripCode(before)), all(/\[\d+(?:[,;–-]\s*\d+)*(?:,\s*с\.\s*\d+(?:[–-]\d+)?)?\]/g, stripCode(after)), v);
  compareExact("ссылка pandoc", all(/\[-?@[^\]\n]+\]/g, stripCode(before)), all(/\[-?@[^\]\n]+\]/g, stripCode(after)), v);
  compareExact("таблица", all(/^[ \t]*\|.*$/gm, stripCode(before)), all(/^[ \t]*\|.*$/gm, stripCode(after)), v);
  compareExact("цитата-блок", all(/^[ \t]*>.*$/gm, stripCode(before)), all(/^[ \t]*>.*$/gm, stripCode(after)), v);
  const quoteNorm = (q: string): string => q.replace(/^["“„«]|["”“»]$/g, "");
  compareExact(
    "цитата",
    all(/«[^«»\n]{1,400}»|"[^"\n]{1,400}"|“[^“”\n]{1,400}”/g, stripCode(before)).map(quoteNorm),
    all(/«[^«»\n]{1,400}»|"[^"\n]{1,400}"|“[^“”\n]{1,400}”/g, stripCode(after)).map(quoteNorm),
    v,
  );
  compareExact("число", numbers(before), numbers(after), v);
  const hb = headingShape(before).join(" ");
  const ha = headingShape(after).join(" ");
  if (hb !== ha) v.push({ kind: "заголовки", detail: `структура заголовков изменилась: [${hb}] → [${ha}]` });

  const issuesBefore = analyze(before, { context }).issues.length;
  const issuesAfter = analyze(after, { context }).issues.length;
  if (issuesAfter > issuesBefore) {
    v.push({ kind: "находки", detail: `находок стало больше: ${issuesBefore} → ${issuesAfter}` });
  }
  return { ok: v.length === 0, violations: v, issuesBefore, issuesAfter };
}

File: src/aiw_ru/validate.py
"""Проверка сохранности: правка прозы не должна портить защищённое содержимое.

Сравнивает исходник и исправленный текст и сообщает, что пропало или изменилось.

Разрешённые скиллом изменения не считаются нарушением: смена прямых кавычек на
«ёлочки», десятичной точки на запятую, Title Case в заголовке на обычный регистр,
удаление параметра utm_source ИИ-инструмента.
"""

from __future__ import annotations

from collections import Counter

from aiw_ru.compat import jsre
from aiw_ru.detect import AI_URL_RE, analyze
from aiw_ru.types import ContextMode, Record


class Violation(Record):
    kind: str
    detail: str


class ValidationResult(Record):
    ok: bool
    violations: list[Violation]
    issues_before: int
    issues_after: int


FRONTMATTER_RE = jsre("^---\\n[\\s\\S]*?\\n---(?:\\n|$)")
FENCE_RE = jsre("^(```|~~~)[^\\n]*\\n[\\s\\S]*?^\\1[^\\n]*$", "gm")
INLINE_CODE_RE = jsre("`[^`\\n]+`", "g")
FORMULA_RE = jsre("\\$\\$[\\s\\S]*?\\$\\$|\\$[^$\\n]+\\$", "g")
URL_RE = jsre('https?:\\/\\/[^\\s)>\\]»"]+', "g")
URL_TAIL_RE = jsre("[?&]$")
LOOSE_URL_RE = jsre("https?:\\/\\/\\S+", "g")
NUMBER_RE = jsre("\\d+(?:[.,]\\d+)*", "g")
CITATION_RE = jsre("\\[\\d+(?:[,;–-]\\s*\\d+)*(?:,\\s*с\\.\\s*\\d+(?:[–-]\\d+)?)?\\]", "g")
PANDOC_RE = jsre("\\[-?@[^\\]\\n]+\\]", "g")
TABLE_RE = jsre("^[ \\t]*\\|.*$", "gm")
BLOCKQUOTE_RE = jsre("^[ \\t]*>.*$", "gm")
QUOTE_RE = jsre('«[^«»\\n]{1,400}»|"[^"\\n]{1,400}"|“[^“”\\n]{1,400}”', "g")
QUOTE_MARKS_RE = jsre('^["“„«]|["”“»]$', "g")
HEADING_SHAPE_RE = jsre("^#{1,6}(?=\\s)", "gm")


def _all(pattern, s: str) -> list[str]:
    return [m.group() for m in pattern.finditer(s)]


def _frontmatter(s: str) -> str:
    m = FRONTMATTER_RE.search(s)
    return m.group() if m else ""


def _strip_code(s: str) -> str:
    return INLINE_CODE_RE.sub("", FENCE_RE.sub("", s))


def _urls(s: str) -> list[str]:
    return [URL_TAIL_RE.sub("", AI_URL_RE.sub("", u), count=1) for u in _all(URL_RE, _strip_code(s))]


def _numbers(s: str) -> list[str]:
    prose = LOOSE_URL_RE.sub("", _strip_code(s))
    return [n.replace(",", ".") for n in _all(NUMBER_RE, prose)]


def _multiset_diff(a: list[str], b: list[str]) -> tuple[list[str], list[str]]:
    count = Counter(a)
    added: list[str] = []
    for x in b:
        if count[x] > 0:
            count[x] -= 1
        else:
            added.append(x)
    missing = [x for x, c in count.items() for _ in range(c)]
    return missing, added


def _compare_exact(kind: str, before: list[str], after: list[str], out: list[Violation]) -> None:
    missing, added = _multiset_diff(before, after)
    out.extend(Violation(kind=kind, detail=f"пропало или изменено: {m[:120]}") for m in missing[:5])
    out.extend(Violation(kind=kind, detail=f"появилось: {a[:120]}") for a in added[:5])


def validate(before: str, after: str, context: ContextMode = "general") -> ValidationResult:
    v: list[Violation] = []
    if _frontmatter(before) != _frontmatter(after):
        v.append(Violation(kind="yaml", detail="YAML-шапка изменена"))
    _compare_exact("код", _all(FENCE_RE, before), _all(FENCE_RE, after), v)
    b, a = _strip_code(before), _strip_code(after)
    _compare_exact("инлайн-код", _all(INLINE_CODE_RE, b), _all(INLINE_CODE_RE, a), v)
    _compare_exact("формула", _all(FORMULA_RE, b), _all(FORMULA_RE, a), v)
    _compare_exact("URL", _urls(before), _urls(after), v)
    _compare_exact("ссылка на литературу", _all(CITATION_RE, b), _all(CITATION_RE, a), v)
    _compare_exact("ссылка pandoc", _all(PANDOC_RE, b), _all(PANDOC_RE, a), v)
    _compare_exact("таблица", _all(TABLE_RE, b), _all(TABLE_RE, a), v)
    _compare_exact("цитата-блок", _all(BLOCKQUOTE_RE, b), _all(BLOCKQUOTE_RE, a), v)
    _compare_exact(
        "цитата",
        [QUOTE_MARKS_RE.sub("", q) for q in _all(QUOTE_RE, b)],
        [QUOTE_MARKS_RE.sub("", q) for q in _all(QUOTE_RE, a)],
        v,
    )
    _compare_exact("число", _numbers(before), _numbers(after), v)
    hb = " ".join(_all(HEADING_SHAPE_RE, b))
    ha = " ".join(_all(HEADING_SHAPE_RE, a))
    if hb != ha:
        v.append(Violation(kind="заголовки", detail=f"структура заголовков изменилась: [{hb}] → [{ha}]"))

    issues_before = len(analyze(before, context).issues)
    issues_after = len(analyze(after, context).issues)
    if issues_after > issues_before:
        v.append(Violation(kind="находки", detail=f"находок стало больше: {issues_before} → {issues_after}"))
    return ValidationResult(ok=not v, violations=v, issues_before=issues_before, issues_after=issues_after)
